Fix canvas showing wrong content after moving frames

CandyFace opened the pull request:
closes #1496 

The PR includes another bugfix, related to when you removed a frame, the canvas would not be updated either.

scribblemaniac commented on core_lib/src/interface/timelinecells.cpp:
```diff
@@ -860,6 +860,11 @@ void TimeLineCells::mouseReleaseEvent(QMouseEvent* event)
+            emit framesMoved();
+
+            mMovingFrames = false;
+        }
```
I don't really like where this is being called from. We're invalidating the cache because of changes in the keyframe position, not directly because of the user's interaction with the timeline. A signal from editor or layer would be better I think.

If we edit Layer::moveSelectedFrames instead, we can call updateFrame for each frame affected (all frames between the old frame position and the new frame position inclusive). This might add a bit of overhead to the dragging interaction, but I think it's worth it for preserving the rest of the frames, and that will be a moot point if/when we merge #1495 since that will only have to be done once after the frames have been dropped.

CandyFace replied:
We could do that but is it necessary? You're not looking at those frames at the given time you're moving, you're only looking at one frame and paint x frames behind and in-front with onion skin enabled, so it should only be necessary to update the current frame and run through the frames affected by onion skin once.

The problem I have with calling update for all affected frames is that we either have to make a new updateFrame function or modify it, as it will otherwise call update onionSkinAround for every affected frame. This should only be necessary for the frame you're looking at, yet we'll be doing it multiple times.

it's easy to clear the cache here and I don't see a big reason to go through hoops to affect specified frames however I'll look into another way of updating, maybe introduce a new cache related flag, to make sure we're not clearing more than needed.

I don't think Layer should have an emitter though, it's just a model, this should exist elsewhere, I agree through that Timeline may not have been the best place, even though it's the closest to the user interface and therefore easiest change without causing side-effects.

scribblemaniac replied:
> You're not looking at those frames at the given time you're moving, you're only looking at one frame and paint x frames behind and in-front with onion skin enabled, so it should only be necessary to update the current frame and run through the frames affected by onion skin once.

No it is not necessary to update the view any frame other than the current one while dragging. The point of what I'm suggesting is not to call it while dragging because they need to be updated before the frame is dropped, but rather because it is the simplest way to implement what you suggest but have not done: update only the affected frames. You can't do this easily in mouseReleaseEvent because you don't know where the frame originated from. You could add logic to handle this and then iterate through the frames inbetween the start and end positions, but why would you when a lot of that could be handled by the changes in #1495?

> as it will otherwise call update onionSkinAround for every affected frame. This should only be necessary for the frame you're looking at, yet we'll be doing it multiple times.

No, this is not enough. At the very least updateOnionSkinsAround needs to be called for the initial and final positions of all selected frames.

> it's easy to clear the cache here and I don't see a big reason to go through hoops to affect specified frames however I'll look into another way of updating, maybe introduce a new cache related flag, to make sure we're not clearing more than needed.

The reason to jump through hoops here is that rendering even a single frame is likely far more expensive than the most inefficient methods of finding the frames which have been modified by a move. I haven't benchmarked this, but intuitively even a few thousand checks in the cache would be nothing compared to the millions of pixels that need to be potentially retrieved from a disk, decoded, allocated, copied, composited, etc. for each and every frame we invalidate and subsequently have to re-render. My philosophy is that we should be using the caches as much as possible, even at the cost of additional code complexity. Playback is arguably our most important functionality, and we need to do more than what we have been doing to assure realtime playback (especially for higher frame rates or lower-powered systems). I'll admit the initial implementation of it has had many problems, but I think we can get this right and I think it will be worth it.

> I don't think Layer should have an emitter though, it's just a model, this should exist elsewhere, I agree through that Timeline may not have been the best place.

Yes you're right, Layer might not be the best choice. I think you understand my issue with where it is currently though. This also ties into why did not suggest doing a single pass at the end of the dragging operation: only the timeline knows if a frame is actively being dragged so I'm not sure how you would do that from anywhere else.

CandyFace commented:
Update:
Alright I went with your simple suggestion, now we simply update selected frames before and after moving. You're right that it will likely always be faster to update moving frames than having to restore a lot of cache, especially if the canvas is big and yes smooth playback should be considered very important.

Because of the rework i've made in swapKeyframe, I hope to see this https://github.com/pencil2d/pencil/issues/1465 fixed too though I haven't been able to reproduce bug myself, the actual logic should be more robust now since a swap can only happen if two keyframes exist where as before one frame could be removed and moved regardless. Of course the result will be the same, the move keyframe logic has just been moved into a new moveKeyFrame function, which replace the moveFrameforward/backwards.

This got me reminded of the move keyframes shortcuts which also did not update cache, so that's also fixed now.

scribblemaniac commented on core_lib/src/structure/layer.cpp:
```diff
@@ -213,72 +213,46 @@ bool Layer::removeKeyFrame(int position)
-        return swapKeyFrames(position, position - 1);
+    if (swapKeyFrames(position, position + offset)) {
+        moved = true;
+    } else {
```
Can you change this if else to use the more standard bracket format please.

CandyFace replied:
Done

scribblemaniac commented on core_lib/src/structure/layer.cpp:
```diff
@@ -213,72 +213,46 @@ bool Layer::removeKeyFrame(int position)
+        if (moveSelectedFrames(offset)) {
+            moved = true;
+        }
+        setFrameSelected(position + offset, false);
```
This approach changes the behavior of the move keyframe shortcut and causes a lot of unwanted behavior when there is already one or more frames selected. Try selecting one or multiple frames and then moving with the cursor on or off a selected frame, with and with out a frame at the destination. There's different behavior in nearly every case now. In v0.6.5 it strictly moves the current keyframe, and I think would should probably just stick with the behavior. I'm fine with changing it to involve the selected frames, but only if we do it in some consistent, sane manner.

CandyFace replied:
Right, that's true.. I like this approach because the movement behavior is consistent with moving frames on the timeline but I can see that there could be some problems with doing it like this... For now I think that saving the selection, deselecting all frames and then re-selecting them afterwards is good enough.

scribblemaniac reviewed (changes requested):
I like the move to a more general moveKeyframe function over moveKeyframeForward/Backward, and I'm alright with swapKeyframe failing if both frames don't exit, but this implementation using a temporary selection for single frame moving opens a whole new can of worms. This is going to need either some discussion of how we want the move keyframe actions to work with a selection, or modifications to make it behave like it did before these changes.

CandyFace commented on core_lib/src/structure/layer.cpp:
```diff
@@ -213,72 +213,46 @@ bool Layer::removeKeyFrame(int position)
-    return swapKeyFrames(position, position + 1);
-}
+    bool moved = false;
+    if (position == 1 && offset < 0) { return false; }
```
Fixed.

CandyFace commented on app/src/actioncommands.cpp:
```diff
@@ -713,9 +713,11 @@ void ActionCommands::moveFrameForward()
     if (layer)
     {
-        if (layer->moveKeyFrameForward(mEditor->currentFrame()))
+        mEditor->notifyFrameMoved();
```
I've reworked this, it made no sense to notify that the cache should be updated from actioncommands (nor anywhere else), instead I think it's up to the model to tell that there are frames to be updated, so that's the approach I've taken.

CandyFace commented:
Ready to be reviewed again.

CandyFace commented on core_lib/src/interface/scribblearea.cpp:
```diff
@@ -197,11 +200,29 @@ void ScribbleArea::updateFrame(int frame)
     update();
 }
 
+void ScribbleArea::updateDirtyFrames()
```
You say that but also say that it's important that we update the cache for the frame before and after moving, so I don't think it's an unnecessary step. By all means if you have a better way of doing, do say so, cause I'm running out of ideas.

This way is imo. much better than making it up to the caller to update the frames it's modifying properly.

The name of the function may not be right but the intention is to make it up to scribblearea to update the cache for the frames that has been modified or can be considered dirty on the currently layer.

scribblemaniac replied:
Let me clarify, I think updating the cache for the frame before and after moving is very important and not unnecessary. What I think is unnecessary is having the QList of frames that you are effectively delaying removing them from the cache. Why not just remove them from the cache right away?

CandyFace replied:
They are being removed at the next frame update, which should happen right after the move, though it could potentially be delayed. it's the least coupled implementation I have come up with and it does the job. I'm not sure how make it happen right away without involving scribblearea in the proces, do you have any ideas? Maybe if we move the qpixmapcache into the layer model, then I would be able to clear it right away.

scribblemaniac replied:
I think part of my confusion came from not paying attention to what was in Layer vs ScribbleArea. I see now that you don't have access to ScribbleArea in Layer and it indeed wouldn't be good to couple that.

As for alternative approaches, the qpixmapcache is for caching rendered+composited frames, so it would not be good to have that in each layer. It also wouldn't work because it needs access to the output of the CanvasPainter, which Layer does not have. It's a tricky problem. I thought maybe you could make a signal and connect the layers to Editor when they are created/loaded, but looking at the code I see that it's really object that's managing the layers, not layermanager, so that wouldn't really work. I have my worries that we might make some change or come across some corner-case where removeCacheForDirtyFrames does not get called, but your solution might be the best thing short of doing some major architectural changes.

CandyFace commented on core_lib/src/interface/scribblearea.h:
```diff
@@ -196,6 +194,13 @@ public slots:
 
 private:
+
+    /** Update cache for dirty keyframes */
```
Ah.. good point.

CandyFace commented on core_lib/src/interface/scribblearea.cpp:
```diff
@@ -175,6 +175,7 @@ void ScribbleArea::setEffect(SETTING e, bool isOn)
 
 void ScribbleArea::updateCurrentFrame()
 {
+    updateDirtyFrames();
```
I don't see why it is necessary to move it, the function won't do anything if there's no keyframes in the list, and if there is.. then we just make sure the cache is not wrong.

scribblemaniac replied:
Well I don't see why it's necessary to call it twice :wink:

CandyFace commented on core_lib/src/structure/layer.cpp:
```diff
@@ -463,9 +472,11 @@ bool Layer::moveSelectedFrames(int offset)
                     if (frame != nullptr)
```
Maybe but I won't make that change here since I have another PR where this function is being reworked.

CandyFace commented on app/src/actioncommands.cpp:
```diff
@@ -713,11 +713,15 @@ void ActionCommands::moveFrameForward()
+        }
+
+        for (int pos : selectedFrames) {
+            layer->setFrameSelected(pos, true);
```
That would alter the functionality of the moveKeyframe method and may cause unwanted behavior, so I don't think that's a good idea. You may not want to always make a deselection and reselect afterwards.

> Additionally, layer stores the ordering of frame selections, but this is not the ordering returned by getSelectedFrames, so it will be lost when this operation is performed. If you move this to layer, you should be able to fix this and perhaps even simply things by manipulating the member variables directly.

Is this important here though? given that we're not dynamically modifying the list and or moving frames at this point, the order has no meaning?

scribblemaniac replied:
> Is this important here though? given that we're not dynamically modifying the list and or moving frames at this point, the order has no meaning?

I'm not sure what you mean by this. What I'm referring to is the order in which the frames are selected, which is stored in `Layer::mSelectedFrames_byLast`. Here's a gif showing how reordering with this deselection/reselection affects the behavior of further frame selections:
![Frame Selection Ordering](https://user-images.githubusercontent.com/3461051/102942105-ae3a9480-4471-11eb-974c-f103d981b9a0.gif)
In this I am clicking frame 5, then ctrl clicking frame 1, then shift clicking frame 3. The second time use the frame forward before the last step. As you can see, it then selects from the first frame selected (frame 5), not the last frame selected (frame 1).

CandyFace replied:
Ah you're right, I used the wrong selection variable. I've simplified the logic now and made it less ugly.

scribblemaniac reviewed (approved):
Thanks for your patience. This is ready to be merged now I think.
